File: middleOfLinkedList/python/middle_of_linked_list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    """

    Complexity
    ----------
    - TC: O(n)
    - SC: O(1)
    """
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        fast = slow = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        return slow


if __name__ == "__main__":
    node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(3)
    node.next.next.next = ListNode(4)

    middle_node = Solution().middleNode(node)
    while middle_node:
        print(middle_node.val)
        middle_node = middle_node.next

File: middleOfLinkedList/c++/middle_of_linked_list.cpp
#include<iostream>

using namespace std;
/**
 * Definition for singly-linked list.
 */
struct ListNode {
     int val;
     ListNode *next;
     ListNode() : val(0), next(nullptr) {}
     ListNode(int x) : val(x), next(nullptr) {}
     ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* fast = head;
        ListNode* slow = head;

        while(fast and fast->next){
            slow = slow->next;
            fast = fast->next->next;

        }
        return slow;
    }
};


//    作者：SunnyYoona
//    来源：CSDN
//    原文：https://blog.csdn.net/sunnyyoona/article/details/42462457
ListNode* createList(int *A, int n){

    if(n <= 0) return NULL;

    ListNode *head = new ListNode(A[0]), *temp = head;

    for(int i = 1; i < n; i++){
        ListNode *input = new ListNode(A[i]);
        temp->next = input;
        temp = temp->next;
    }

    return head;

}

int main() {

    int A[] = {1,2,4,7,9};
    ListNode* head = createList(A,5);

    Solution ans;
    ListNode* answer = ans.middleNode(head);

    ListNode *p = answer;
    while(p){
        cout<<p->val<<" ";
        p = p->next;
    }
    cout<<endl;

    return 0;
}